### ISSUE TITLE
`Style/RedundantCondition` is doing unsafe corrections on methods that end in `?`

### ISSUE THREAD

**Issue by a-lavis** [rubocop/rubocop#13931]
> It doesn't seem like this is fixed in 1.73.1 for me:
> 
> Given the following `.rubocop.yml`:
> ```yml
> AllCops:
>   NewCops: enable
>   TargetRubyVersion: 3.4
> ```
> and the following `foo.rb`:
> ```ruby
> # frozen_string_literal: true
> 
> def foo? = 123
> 
> puts foo? ? true : false
> ```
> 
> If you run `ruby foo.rb`, you get `true`.
> 
> Then if you run `rubocop -a foo.rb`, it changes it to:
> ```ruby
> # frozen_string_literal: true
> 
> def foo? = 123
> 
> puts foo? || false
> ```
> 
> Then, if you run `ruby foo.rb`, you instead get `123`.
> 
> So it seems like this is still not a safe autocorrect.
> 
> `rubocop -V`:
> ```
> 1.73.1 (using Parser 3.3.7.1, rubocop-ast 1.38.1, analyzing as Ruby 3.4, running on ruby 3.4.2) [arm64-darwin24]
> ``` 

 _Originally posted by @a-lavis in [#13913](https://github.com/rubocop/rubocop/issues/13913#issuecomment-2688336623)_

Decided to pull this comment out into its own issue since in retrospect it feels like it might be a separate issue/discussion - but feel free to close this issue if we should be using the original issue instead!

**Comment by bbatsov** [comment]
That's intentional. RuboCop can't know which predicate is returning `true` or `false`, or something else, but it assumes that predicates are well-behaved by default and people can explicitly list predicates they want to treat differently in the config.

We've opted for a solution that's not perfectly safe, but makes the cop more useful in general.

**Comment by a-lavis** [comment]
Ah okay - given this:

> We've opted for a solution that's not perfectly safe, but makes the cop more useful in general.

I thought it would be an unsafe autocorrect instead of a safe one.

If you all think it makes more sense as a safe autocorrect given that it can be configured, feel free to close this issue!

Thank you!

**Comment by a-lavis** [comment]
Was thinking about it more over the weekend, and there are lots of examples of cops that have unsafe autocorrects like this one.

For example [`Style/NumericPredicate`](https://docs.rubocop.org/rubocop/cops_style.html#stylenumericpredicate) unsafe autocorrects from `foo == 0` to `foo.zero?`; while this is not perfectly safe, its probably fine because `foo` is probably a number.
I think its appropriate this is an unsafe autocorrect, because it makes the cop more useful but its not necessarily safe in all cases.

There are even cops that are more safe than `Style/RedundantCondition` which are marked as unsafe.  For example [`Style/CommentedKeyword`](https://docs.rubocop.org/rubocop/cops_style.html#stylecommentedkeyword), which, since it only interacts with comments, cannot semantically change the program (whereas `Style/RedundantCondition` can).

---

I guess the question is: what makes an autocorrection safe vs unsafe?

According to this page: https://docs.rubocop.org/rubocop/usage/autocorrect.html

https://github.com/rubocop/rubocop/blob/0037549434928f40c0862dab1d15f88fe9e2977a/docs/modules/ROOT/pages/usage/autocorrect.adoc?plain=1#L17-L19

> - Some automatic corrections might change (slightly) the semantics of the code,
meaning they'd produce code that's mostly equivalent to the original code, but
not 100% equivalent. We call such autocorrect behavior "unsafe".

https://github.com/rubocop/rubocop/blob/0037549434928f40c0862dab1d15f88fe9e2977a/docs/modules/ROOT/pages/usage/autocorrect.adoc?plain=1#L32-L41

> In RuboCop 0.60, we began to annotate cops as `Safe` or not safe. The definition of
> safety is that the cop doesn't generate false positives. On top of that there's `SafeAutoCorrect`
> that might be set to `false` in cases where only the autocorrect performed by a cop
> is unsafe, but that the offense detection logic is safe. To sum it up:
>
> * Safe (`true/false`) - indicates whether the cop can yield false positives (by
design) or not.
> * SafeAutoCorrect (`true/false`) - indicates whether the autocorrect a cop
does is safe (equivalent) by design. If a cop is unsafe its autocorrect automatically
becomes unsafe as well.

According to this definition:
> The definition of safety is that the cop doesn't generate false positives.

`Style/RedundantCondition` would be considered unsafe, because it can generate false positives.  Maybe this definition is out of date though.

**Comment by jdelStrother** [comment]
I don't think this is limited to methods that end in `?`, unless I'm looking at a completely separate issue.

eg this - 

```ruby
    def prices(all: false)
      active = all ? true : nil
      Stripe::Price.list(active: active)
    end
```

gets autocorrected to:

```ruby
    def prices(all: false)
      active = all || nil
      Stripe::Price.list(active: active)
    end
```

Which I guess is fine if you're calling prices with the expected true/false, but it does place that onus onto the caller of the method rather than just handling it with a simple ternary.

(Plus, subjectively, `all ? true : nil` is far clearer to me than `all || nil`, even _if_ you know that `all` is either true or false)

**Comment by jdelStrother** [comment]
Ahh, thanks - I didn't try the version from main.

**Comment by Earlopain** [comment]
That PR merge preceeded your comment anyways, no worries
